lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1417 @@
+    )
+
+    $filter = "Name='{0}'" -f $ServiceName
+
+    $service = Get-WmiObject -Namespace "root\cimv2" -Class Win32_Service -Filter $filter
+    if($service) {
+        Write-JujuLog "Service $ServiceName is already created."
+        return $true
+    }
+
+    $serviceFileName = "OpenStackService.exe"
+    if(!(Test-Path "$SERVICE_DIR\$serviceFileName")) {
+        Copy-Item "$FILES_DIR\$serviceFileName" "$SERVICE_DIR\$serviceFileName"
+    }
+
+    New-Service -Name "$ServiceName" -DisplayName "$ServiceName" -Description "$ServiceDescription" -StartupType "Manual" `
+                -BinaryPathName "$SERVICE_DIR\$serviceFileName $ServiceName $ServiceExecutable --config-file $ServiceConfig" `
+
+    if((Get-Service -Name $ServiceName).Status -eq "Running") {
+        Stop-Service $ServiceName
+    }
+
+    Set-ServiceLogon -Services @($ServiceName) -UserName $ServiceUser -Password $ServicePassword
+}
+
+
+function Watch-ServiceStatus {
+    Param(
+        [string]$ServiceName,
+        [int]$IntervalSeconds
+    )
+
+    $count = 0
+    while ($count -lt $IntervalSeconds) {
+        if ((Get-Service -Name $ServiceName).Status -ne "Running") {
+            $count += 1
+            Start-Sleep -Seconds 1
+            continue
+        }
+        return
+    }
+    Throw "$ServiceName has errors. Please check the logs."
+}
+
+
+function Get-JujuVMSwitchName {
+    $VMswitchName = Get-JujuCharmConfig -Scope "vmswitch-name"
+    if (!$VMswitchName){
+        return "br100"
+    }
+    return $VMswitchName
+}
+
+
+function Get-InterfaceFromConfig {
+    Param (
+        [string]$ConfigOption="data-port",
+        [switch]$MustFindAdapter=$false
+    )
+
+    $nic = $null
+    $DataInterfaceFromConfig = Get-JujuCharmConfig -Scope $ConfigOption
+    Write-JujuInfo "Looking for $DataInterfaceFromConfig"
+    if (!$DataInterfaceFromConfig){
+        if($MustFindAdapter) {
+            Throw "No data-port was specified"
+        }
+        return $null
+    }
+    $byMac = @()
+    $byName = @()
+    $macregex = "^([a-f-A-F0-9]{2}:){5}([a-fA-F0-9]{2})$"
+    foreach ($i in $DataInterfaceFromConfig.Split()){
+        if ($i -match $macregex){
+            $byMac += $i.Replace(":", "-")
+        }else{
+            $byName += $i
+        }
+    }
+    if ($byMac.Length){
+        $nicByMac = Get-NetAdapter | Where-Object { $_.MacAddress -in $byMac -and $_.DriverFileName -ne "vmswitch.sys" }
+    }
+    if ($byName.Length){
+        $nicByName = Get-NetAdapter | Where-Object { $_.Name -in $byName }
+    }
+    if ($nicByMac -ne $null -and $nicByMac.GetType() -ne [System.Array]){
+        $nicByMac = @($nicByMac)
+    }
+    if ($nicByName -ne $null -and $nicByName.GetType() -ne [System.Array]){
+        $nicByName = @($nicByName)
+    }
+    $ret = $nicByMac + $nicByName
+    if ($ret.Length -eq 0 -and $MustFindAdapter){
+        Throw "Could not find network adapters"
+    }
+    return $ret
+}
+
+
+function Get-DataPortFromDataNetwork {
```

Comment:
Similar to what we do [here](https://github.com/cloudbase/nova-hyperv-charm/blob/master/lib/Modules/OVSCharmUtils/OVSCharmUtils.psm1#L83). You decide if it's exactly what you need and you can import the function from there and remove it here.